### DISCUSSION TITLE
动态配置数据有什么好的方法

### DISCUSSION THREAD

**Discussion by jamiesun** [IrineSistiana/mosdns#59] [Q&A]
`
  - tag: query_myips
    type: query_matcher
    args:
      client_ip:
        - ./conf/myips
`

目前在匹配规则中，client_ip 都是定义好的数据文件，服务启动时加载，但是如果这个 client_ip 数据列表是动态的， 有什么好的办法解决

**Comment by IrineSistiana** [comment]
目前只能 restart.

**Reply by jamiesun** [reply]
我试着做了下改造， 引入nanomsg 的发布订阅机制， 在mosdns 上订阅远程服务器上的数据， 当远程服务器IP数据更新， 就会发布数据， 这样所有mosdns节点都可以收到数据然后更新内存里的规则。

不过目前mosdns里加载到内存里的插件规则数据还没有方便调用的管理方法， 比如update， remove

我正在试着对 query_match 加入改造

```
type queryMatcher struct {
	*handler.BP
	args *Args

	matcherGroup []handler.Matcher
}

func (qm *queryMatcher) updateMatcher(...){
 ...
}

```

**Reply by jamiesun** [reply]
我想复杂了，其实只要在函数里reload 配置就好了

func (qm *queryMatcher) updateMatcher(){
 // reload config file of args
}

**Reply by IrineSistiana** [reply]
如果只是订阅 ip 变化的话，我觉得单独搞一个插件也许更好一些？🤔

`queryMatcher` 只是为了配置文件方便才堆在一起的。

**Reply by jamiesun** [reply]
目前也是在探索阶段， 实际使用时很多配置希望动态去更新， 在管理系统中实现规则下发

**Reply by aa51513** [reply]
> 我试着做了下改造， 引入nanomsg 的发布订阅机制， 在mosdns 上订阅远程服务器上的数据， 当远程服务器IP数据更新， 就会发布数据， 这样所有mosdns节点都可以收到数据然后更新内存里的规则。
> 
> 不过目前mosdns里加载到内存里的插件规则数据还没有方便调用的管理方法， 比如update， remove
> 
> 我正在试着对 query_match 加入改造
> 
> ```
> type queryMatcher struct {
> 	*handler.BP
> 	args *Args
> 
> 	matcherGroup []handler.Matcher
> }
> 
> func (qm *queryMatcher) updateMatcher(...){
>  ...
> }
> ```

弄好了吗，大佬？我也有这个需要
